src/hotspot/os/aix/os_perf_aix.cpp

```diff
@@ -22,519 +22,382 @@
  */
-static OSReturn get_total_ticks(int which_logical_cpu, CPUPerfTicks* pticks) {
-  return OS_ERR;
-}
+static void calculate_updated_load(CPUPerfTicks* update, CPUPerfTicks* prev, double* load) {
+  CPUPerfTicks diff;
 
-/** read user and system ticks from a named procfile, assumed to be in 'stat' format then. */
-static int read_ticks(const char* procfile, uint64_t* userTicks, uint64_t* systemTicks) {
-  return read_statdata(procfile, "%*c %*d %*d %*d %*d %*d %*u %*u %*u %*u %*u " UINT64_FORMAT " " UINT64_FORMAT,
-    userTicks, systemTicks);
-}
+  if (update && prev && load) {
+    diff.user = update->user - prev->user;
+    diff.sys  = update->sys  - prev->sys;
+    diff.idle = update->idle - prev->idle;
+    diff.wait = update->wait - prev->wait;
 
-/**
- * Return the number of ticks spent in any of the processes belonging
- * to the JVM on any CPU.
- */
-static OSReturn get_jvm_ticks(CPUPerfTicks* pticks) {
-  return OS_ERR;
+    *load = 1.0 - diff.idle/(diff.sys + diff.user + diff.idle + diff.wait);
+  }
 }
 
 /**
- * Return the load of the CPU as a double. 1.0 means the CPU process uses all
- * available time for user or system processes, 0.0 means the CPU uses all time
- * being idle.
- *
- * Returns a negative value if there is a problem in determining the CPU load.
+ * Look up lcpu names for later re-use.
  */
-static double get_cpu_load(int which_logical_cpu, CPUPerfCounters* counters, double* pkernelLoad, CpuLoadTarget target) {
-  uint64_t udiff, kdiff, tdiff;
-  CPUPerfTicks* pticks;
-  CPUPerfTicks  tmp;
-  double user_load;
+static bool populate_lcpu_names(int ncpus, perfstat_id_t* lcpu_names) {
+  perfstat_cpu_t* all_lcpu_stats;
+  perfstat_cpu_t* lcpu_stats;
+  perfstat_id_t   name_holder;
 
-  *pkernelLoad = 0.0;
+  assert(lcpu_names, "Names pointer NULL");
+  assert(initialize_libperfstat(), "perfstat lib not available");
 
-  if (target == CPU_LOAD_VM_ONLY) {
-    pticks = &counters->jvmTicks;
-  } else if (-1 == which_logical_cpu) {
-    pticks = &counters->cpus[counters->nProcs];
-  } else {
-    pticks = &counters->cpus[which_logical_cpu];
-  }
+  strncpy(name_holder.name, FIRST_CPU, IDENTIFIER_LENGTH);
 
-  tmp = *pticks;
+  all_lcpu_stats = NEW_RESOURCE_ARRAY(perfstat_cpu_t, ncpus);
 
-  if (target == CPU_LOAD_VM_ONLY) {
-    if (get_jvm_ticks(pticks) != OS_OK) {
-      return -1.0;
-    }
-  } else if (get_total_ticks(which_logical_cpu, pticks) != OS_OK) {
-    return -1.0;
-  }
-
-  // seems like we sometimes end up with less kernel ticks when
-  // reading /proc/self/stat a second time, timing issue between cpus?
-  if (pticks->usedKernel < tmp.usedKernel) {
-    kdiff = 0;
-  } else {
-    kdiff = pticks->usedKernel - tmp.usedKernel;
+  // populate cpu_stats && check that the expected number of records have been populated
+  if (ncpus != libperfstat::perfstat_cpu(&name_holder, all_lcpu_stats, sizeof(perfstat_cpu_t), ncpus)) {
+    FREE_RESOURCE_ARRAY(perfstat_cpu_t, all_lcpu_stats, ncpus);
```

Comment:
ResourceArray are stack-based arenas. No need to free them manually (freeing is often a noop anyway unless the allocation is a the top of the arena). Very similar to those AutoReleasePools in Objective-C, if you know that.

The way to use them is don't release them. If those allocations don't escape the function (so, you don't return RA memory from the function), you can set a ResourceMark at the start of the function. That is an RAII object which will clear all RA objects allocated in and under this function upon return.

Arguably, you also could just use malloc and free here (NEW_C_HEAP_ARRAY and FREE_C_HEAP_ARRAY)

Reply:
This is interesting, and definitely good to know. I have removed all FREE_RESOURCE_ARRAY calls and added a ResourceMark to the top of all procedures which use RESOURCE_ARRAYs.